### PR TITLE
[Spark] Add BackfillBatchIterator for Row Tracking Backfill

### PR DESCRIPTION
**Pull request by longvu-db**
#### Which Delta project/connector is this regarding?

- [X] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description
In this PR, we add `BackfillBatchIterator` which contains the core logic for selecting files for the WIP Row Tracking Backfill operation, creating iterators which batches files together. The Row Tracking Backfill operation will have multiple commits, each commit adds `baseRowId` to a batch of files. More details can be found in the [Design Doc](https://docs.google.com/document/d/1ji3zIWURSz_qugpRHjIV_2BUZPVKxYMiEFaDORt_ULA/edit#heading=h.8al9qhd83yov).

## How was this patch tested?
Added new Test Suite.

## Does this PR introduce _any_ user-facing changes?
No.

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/backfill/BackfillBatchIterator.scala [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+  private val addFileItr = new Iterator[AddFile] {
+    private val underlying = collectedFiles.iterator
+
+    override def hasNext: Boolean = underlying.hasNext
+
+    override def next(): AddFile = underlying.next()
+  }
+
+  private val sourcePeekableItr = addFileItr.buffered
```
Comment (johanl-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/commands/backfill/BackfillBatchIterator.scala:48]: nit: do we actually need this?
We're not using `sourcePeekableItr.head` so directly using `addFileItr` should be enough

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/backfill/BackfillBatchIterator.scala [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  private val buffer: ListBuffer[AddFile] = ListBuffer.empty[AddFile]
+
+  private var lastBin: Seq[AddFile] = Nil
+
+  private def createBin(): Seq[AddFile] = {
+    val taskLevelPermitAllocator = tracker.createTaskLevelPermitAllocator()
+
+    while (sourcePeekableItr.hasNext) {
+      // Acquire permit to fetch the next file from source.
+      taskLevelPermitAllocator.acquirePermit()
```
Comment (johanl-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/commands/backfill/BackfillBatchIterator.scala:59]: When are we (or will we) release the permits?

**Review by johanl-db — APPROVED** [review]
Minor comments, looks good to me

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/backfill/BackfillBatchIterator.scala (hunk 2 above)
Comment (longvu-db) [reply spark/src/main/scala/org/apache/spark/sql/delta/commands/backfill/BackfillBatchIterator.scala:59]: @johanl-db Yes we will, the plan is after we attempt to execute the backfill batch, we will release all the permits for the files materialized by the batch.

**Pull request merged**